proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,302 @@
+cluster. There are proposals of more aggressive block production reduction to
+save the cluster. For example, a leader could produce only one block in four
+consecutive slots allocated to it.
+
+However, this only solves the problem in specific type of outage, and it seems
+risky to aggressively reduce block production, so we are not proceeding with
+this proposal for now.
+
+## Detailed Design
+
+The new protocol tries to make all restarting validators get the same data
+blocks and the same set of last votes, then they will almost certainly make the 
+same decision on the canonical fork and proceed.
+
+A new command line arg will be added. When the cluster is in need
+of a restart, we assume validators holding at least `RESTART_STAKE_THRESHOLD`
+percentage of stakes will restart with this arg. Then the following steps
+will happen:
+
+1. The validator boots into the silent repair phase, it will not make new 
+blocks or change its votes. The validator propagates its local voted fork
+information to all other validators in restart.
+
+2. While counting local vote information from all others in restart, the
+validator repairs all blocks which could potentially have been optimistically
+confirmed.
+
+3. After repair is complete, the validator counts votes on each fork and
+sends out local heaviest fork.
+
+4. Each validator counts if enough nodes can agree on one block (same slot and
+hash) to restart from:
+
+  1. If yes, proceed and restart
+
+  2. If no, print out what it thinks is wrong, halt and wait for human
+
+See each step explained in details below.
+
+### 1. Gossip last vote before the restart and ancestors on that fork
+
+The main goal of this step is to propagate the locally selected fork to all
+others in restart.
+
+We use a new Gossip message `LastVotedForkSlots`, its fields are:
+
+- `last_voted_slot`: `u64` the slot last voted, this also serves as last_slot
+for the bit vector.
+- `last_voted_hash`: `Hash` the bank hash of the slot last voted slot.
+- `ancestors`: `BitVec<u8>` compressed bit vector representing the slots on
+sender's last voted fork. the most significant bit is always
+`last_voted_slot`, least significant bit is `last_voted_slot-81000`.
+
+The number of ancestor slots sent is hard coded at 81000, because that's
+400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9 
+hours. If a validator restarts after 9 hours past the outage, it cannot join 
+the restart this way. If enough validators failed to restart within 9 hours, 
+then fallback to the manual, interactive cluster restart method.
+
+When a validator enters restart, it uses ephemeral shred version to avoid
+interfering with those outside the restart. There is slight chance that 
+the ephemeral shred version would collide with the shred version after the
+silent repair phase, but even if this rare case occurred, we plan to flush the 
+CRDS table on successful restart, so gossip messages used in restart will be 
```

Comment:
When in normal gossip mode, I feel like we also shouldn't pull the `LastVotedForkSlots` and `Heaviest` messages and filter them out from a push. Otherwise these messages could end up floating around in gossip anyway

Reply:
Sure we can do that, added.

Reply:
First of all I feel we should select repair peers only from validators in restart, because at this point you don't know what the status of non-restarted validators are, they could be buried in Turbine/Gossip/random traffic so they can't effectively answer your repair requests.

Chatted a bit with Ashwin, I feel like to not pollute normal Gossip, there are two options:
1. Use "silent restart shred version", but this means you have to add in a hack where during repairs you accept shreds in original shred version.
2. Make validators in restart drop normal Gossip messages, and make validators not in restart drop restart Gossip messages (keep the ClusterInfo messages, of course), then we don't need the "silent restart shred version".

Reply:
2 seems nice if it'll work. Gossip has proven itself to be incredibly resilient so I think it likely will

Reply:
@behzadnouri Do you think option 2 would work?

Reply:
Hmm, option 2 might not work too well when the number of restarted validators is small. Because when a restarted validator try to push LastVotedForkSlots to others, the peers it selected might all be non-restarted validators which drop this message. Of course, a newly restarted validator can pull the messages, but it needs to know which restarted peer to pull from.
Basically with option 2 we still need to know who's restarted when pushing and pulling to make gossip efficient. But this involves more hacks in gossip code.

Reply:
Thought about this more, since this is outage handling, I'd prefer the solution to be less intrusive so it doesn't interfere with normal logic.
Read the code a bit more, I think manipulating shred_version is still the less intrusive option. We can use only the silent_repair_shred_version on Gossip and uses the original shred_version elsewhere so that repair would just work.
This would mean validators not in restart are able to send over shreds, it may consume more memory but I think it's not worth putting in additional hack to block, because if any validator is malicious then it's easy to send shreds belonging to any shred_version.
So I'd still stick with option 1. We can maybe filter out the restart Gossip messages on validators not in restart if we want, even if we don't, it's just doubling the resource needed for EpochSlots, so it should be fine as well.

Reply:
sg, I'm on board with that approach. I agree it should be much less intrusive to the existing Labs validator code base